linkerd/proxy/discover/src/from_resolve.rs

```diff
@@ -163,50 +185,112 @@ where
     fn addr(n: u8) -> SocketAddr {
         SocketAddr::from(([10, 1, 1, n], PORT))
     }
 
-    #[tokio::test]
+    #[tokio::test(flavor = "current_thread")]
     async fn reset() {
-        tokio::pin! {
-            let stream = stream! {
-                yield Ok::<_, Infallible>(Update::Add((1..=2).map(|n| (addr(n), n)).collect()));
-                yield Ok(Update::Reset((2..=4).map(|n| (addr(n), n)).collect()));
-            };
-        }
-        let mut disco = Discover::new(stream);
-
-        for i in 1..=2 {
-            match disco.next().await.unwrap().unwrap() {
-                Change::Remove(_) => panic!("Unexpectd Remove"),
-                Change::Insert(a, n) => {
-                    assert_eq!(n, i);
-                    assert_eq!(addr(i), a);
-                }
-            }
-        }
-        match disco.next().await.unwrap().unwrap() {
-            Change::Remove(a) => assert_eq!(a, addr(1)),
-            change => panic!("Unexpected change: {:?}", change),
+        let (tx, rx) = tokio::sync::mpsc::channel(1);
+        let mut disco = Discover::new(ReceiverStream::new(rx));
+
+        // Use reset to set a new state with 3 addresses.
+        tx.try_send(Ok::<_, Infallible>(Update::Reset(
+            (1..=3).map(|n| (addr(n), n)).collect(),
+        )))
+        .expect("must send");
+        for i in 1..=3 {
+            assert!(matches!(
+                disco.try_next().await,
+                Ok(Some(Change::Insert(sa, n))) if sa == addr(i) && n == i
+            ));
         }
+
+        // Reset to a new state with 3 addresses, one of which is unchanged, one of which is
+        // changed, and one of which is added.
+        tx.try_send(Ok(Update::Reset(
+            // Restore the original `2` value. We shouldn't see an update for it.
+            Some((addr(2), 2))
+                .into_iter()
+                // Set a new value for `3`. and add a 4th as well.
+                .chain((3..=4).map(|n| (addr(n), n + 1)))
+                .collect(),
+        )))
+        .expect("must send");
+        // The first address is removed now.
+        assert!(matches!(
+            disco.try_next().await,
+            Ok(Some(Change::Remove(a))) if a == addr(1)
+        ));
+        // Then process the changed and new addresses.
         for i in 3..=4 {
-            match disco.next().await.unwrap().unwrap() {
-                Change::Remove(_) => panic!("Unexpectd Remove"),
-                Change::Insert(a, n) => {
-                    assert_eq!(n, i);
-                    assert_eq!(addr(i), a);
-                }
-            }
+            assert!(matches!(
+                disco.try_next().await,
+                Ok(Some(Change::Insert(sa, n))) if sa == addr(i) && n == i + 1
+            ));
         }
+
+        // No more updates.
+        drop(tx);
         assert!(disco.next().await.is_none());
     }
+
+    #[tokio::test(flavor = "current_thread")]
+    async fn dedupe() {
```

Comment:
tiny, silly nit: generally i think i've seen this spelt "dedup" (it's not spelled "dupelicate"...)